csp/random.py

```diff
@@ -0,0 +1,167 @@
+import numpy as np
+from datetime import timedelta
+from typing import TypeVar
+
+import csp
+import csp.stats
+from csp import ts
+from csp.typing import Numpy1DArray, NumpyNDArray
+
+__all__ = ("poisson_timer", "brownian_increments", "brownian_motion", "brownian_motion_1d")
+
+
+T = TypeVar("T")
+
+
+@csp.node
+def poisson_timer(rate: ts[float], seed: object, value: "~T" = True) -> ts["T"]:
+    """Generate events according to a Poisson process with time-varying rate.
+    For a fixed-interval timer see csp.timer.
+
+    Args:
+        rate: The rate of the Poisson process (per second), must be non-negative
+        seed: The seed for the numpy random Generator. Can be anything accepted by np.random.default_rng
+        value: The value to tick when there are events (similar to csp.timer)
+    """
+    with csp.alarms():
+        event = csp.alarm(bool)
+
+    with csp.state():
+        s_rng = np.random.default_rng(seed)
+        s_scheduled_event = None
+
+    if csp.ticked(rate):
+        if rate < 0:
+            raise ValueError(f"{csp.now()}: rate must be non-negative")
+        if s_scheduled_event:
+            csp.cancel_alarm(event, s_scheduled_event)
+        if rate > 0:
+            seconds = s_rng.exponential(1.0 / rate)
+            s_scheduled_event = csp.schedule_alarm(event, timedelta(seconds=seconds), True)
+
+    if csp.ticked(event):
+        seconds = s_rng.exponential(1.0 / rate)
+        s_scheduled_event = csp.schedule_alarm(event, timedelta(seconds=seconds), True)
+        return value
+
+
+def _make_brownian_increment(t, s_rng, drift, s_cov_decomp):
+    # Make the brownian increment as efficiently as possible
+    values = s_rng.normal(scale=np.sqrt(t), size=drift.size)
+    np.dot(s_cov_decomp, values, out=values)
+    np.add(drift * t, values, out=values)
+    return values
+
+
+def _matrix_decomposition(matrix, now):
+    # Split matrix decomposition into its own function to help with profiling and make it easier to replace
```

Comment:
I chose svd because it was the default choice for [np.random.Generator.multivariate_normal](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.multivariate_normal.html). I don't really see the need for other methods (as I expect covariance matrix not to change much, but we can always add the eigenvalue based method down the road).